File: billing_system/frontend/flutter_application/lib/models/customer.dart
/// Customer model matching the Flask API response.
class Customer {
  final String id;
  final String name;
  final String phone;
  final String address;
  final String email;
  final String area;
  final String gstin;
  final double creditLimit;
  final double balance;

  const Customer({
    required this.id,
    required this.name,
    required this.phone,
    required this.address,
    this.email = '',
    this.area = '',
    this.gstin = '',
    this.creditLimit = 0,
    this.balance = 0,
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id:          json['id'].toString(),
      name:        json['name'] as String,
      phone:       json['phone'] as String? ?? '',
      address:     json['address'] as String? ?? '',
      email:       json['email'] as String? ?? '',
      area:        json['area'] as String? ?? '',
      gstin:       json['gstin'] as String? ?? '',
      creditLimit: (json['credit_limit'] as num?)?.toDouble() ?? 0,
      balance:     (json['balance'] as num?)?.toDouble() ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'phone': phone,
        'address': address,
        'email': email,
        'area': area,
        'gstin': gstin,
        'credit_limit': creditLimit,
        'balance': balance,
      };

  /// Walk-in customer singleton used as default selection.
  static const Customer walkIn = Customer(
    id: '00000000-0000-0000-0000-000000000000',
    name: 'Walk-in Customer',
    phone: '',
    address: '',
    area: 'General',
  );

  @override
  String toString() => 'Customer($id, $name)';
}

File: billing_system/backend/models/customer.py
"""
Customer model.
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class Customer:
    id: str            # UUID from Supabase
    name: str
    phone: str
    address: str
    email: Optional[str] = field(default="")
    area: Optional[str] = field(default="")
    gstin: Optional[str] = field(default="")
    credit_limit: float = 0.0
    balance: float = 0.0

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "phone": self.phone,
            "address": self.address,
            "email": self.email,
            "area": self.area,
            "gstin": self.gstin,
            "credit_limit": self.credit_limit,
            "balance": self.balance,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Customer":
        return cls(
            id=str(data["id"]),
            name=data["name"],
            phone=data.get("phone", ""),
            address=data.get("address", ""),
            email=data.get("email", ""),
            area=data.get("area", ""),
            gstin=data.get("gstin", ""),
            credit_limit=float(data.get("credit_limit", 0)),
            balance=float(data.get("balance", 0)),
        )
